### ISSUE TITLE
GSNR computation incorrect?

### ISSUE THREAD

**Issue by andres-fr** [f-dangel/cockpit#28]
I have a question regarding the scaling of the GSNR computation as provided in Cockpit

## Description

In the following gist, I've written 3 different ways of computing the GSNR as defined in [this paper](https://openreview.net/forum?id=HyevIJStwH), the same one cited in the Cockpit paper:

1. Extracting the variance directly from the sample-wise gradients, via BackPACK `BatchGrad()` extension (bad, memory inefficient)
2. Using the `Variance()` BackPACK extension (efficient & straightforward)
3. Using a replacement formula for the variance to avoid sample-wise grad squaring:  `Var(X) = E[X²] - E[X]²` (likely faster, less straightforward)

https://gist.github.com/andres-fr/db9d0ba31d1502df62a09d382e504d1e

Specifically, the first and second reflect the mathematical definition from the paper:
```
GSNR_paper = grad_squared / Var(gradients)
```

And this is the implementation of the replacement formula that I found to be equivalent:
```
GSNR_gist = grad_squared / ( (sum_grad_squared / batch_size) - (grad_squared / (batch_size**2)) + epsilon)
```

My issue/question is when I compare the last expression to the one in the [Cockpit source code](https://github.com/f-dangel/cockpit/blob/af91391ddab2a8aef85905b081ccf67d94c1a0e5/cockpit/quantities/mean_gsnr.py#L82), a clear difference can be noticed in the batch scaling:

```
GSNR_cockpit = grad_squared / (batch_size * sum_grad_squared - grad_squared + epsilon)
```

When I used this last expression in my gist, it led to results clearly different from the other approaches.
Am I missing something here? Is cockpit doing some further scaling under the hood? Or are these results scaled differently from the definition for some reason?

----

## Steps to Reproduce

The gist works on CPU and CUDA. For easier reproduction, the gist depends on PyTorch and BackPACK only. The script docstring includes steps to install a minimal working conda environment. Then, simply run the script.

## Source or Possible Fix

The [BackPACK docs](https://docs.backpack.pt/en/master/extensions.html#backpack.extensions.Variance) already warned about scaling issues with the Variance, which may even involve the loss function, so I wonder if this is a bug/feature.

Intuitively, I'd say it is good if the quantity matches the definition from the paper, but again, implicit scaling may be getting in the way?

**Comment by f-dangel** [comment]
Hi, thanks for the detailed report!

I didn't have time to check the details yet, but the cause of this issue is probably that BackPACK's quantities have certain scalings (details below). I am not sure yet if the bug is on your or our side, but either way it would be great to figure this out. Maybe you can double-check the GSNR computation with a for-loop and `torch.autograd`, as done by the [test](https://github.com/f-dangel/cockpit/blob/main/tests/test_quantities/test_mean_gsnr.py#L57-L81)?

Details on BackPACK quantities: Assume the loss is $L = 1 / N \sum_{n=1}^{N} \ell_n$, and the gradient is $\nabla L = 1 / N \sum_{n=1}^{N} \nabla \ell_n$ with $g_n = \nabla \ell_n$ the gradient of sample $n$. Then:
- BackPACK's `BatchGrad` computes $\\{1 / N g_n\\}_{n=1}^N$  (see the [doc](https://docs.backpack.pt/en/master/extensions.html#backpack.extensions.BatchGrad))
- BackPACK's `SumGradSquared` computes the sum of $\\{(1 / N g_n)^2\\}_{n=1}^N$  (see the [doc](https://docs.backpack.pt/en/master/extensions.html#backpack.extensions.SumGradSquared))
- BackPACK's `Variance` computes the variance of the vectors $\\{1 / N g_n\\}_{n=1}^N$  (see the [doc](https://docs.backpack.pt/en/master/extensions.html#backpack.extensions.Variance))

Hope this helps! Keep me posted.

Best,
Felix

**Comment by andres-fr** [comment]
Dear Felix,

I agree with you that the issue is with the batch-related scaling. But I don't agree that the issue is related to an autograd/for loop implementation, for the following reason: In the test you linked, the same formula as Cockpit is being used, and my question is whether the formula itself is correct. So even if the test passes against that formula, that doesn't answer the question.

So the way I thought would be most meaningful to identify the issue, would be to take your definitions, the GSNR definition, and do some algebra to identify the effects of different scaling being involved, both in the Cockpit formula as in mine. Below are the results. The conclusion is that Cockpit is right and my gist is wrong, so no changes are needed and I'm closing this issue.

![gsnr_backpack](https://user-images.githubusercontent.com/4235319/202747581-a2772e69-2a53-48ae-a076-249d44eef65c.png)


And [here](https://github.com/f-dangel/cockpit/files/10042814/gsnr_backpack.txt)  is the Latex file for future reference.

**Comment by f-dangel** [comment]
Thanks for the prompt response. Do you think there is a better space to  document the scaling of these quantities, other than the docstring of the respective BackPACK extension?

**Comment by andres-fr** [comment]
Taking a step back, the main pain point here is IMO that the scaling depends on the loss function, which is out of Cockpit's control.

The clean fix would be to have access to this information and scale the quantities accordingly, so that there is a unique, inconditional formula for the quantities.

But this would introduce more issues: either the user must provide this to Cockpit (leading to potential inconsistencies and overhead) or the code does some hack to figure this out itself (I'm not sure this is possible in a maintainable+clean way, do you?)

So what is left is, as you suggest, documentation. The general theme here seems to be that the literature uses `g` as quantity for the gradients, but in most cases backpack will use `g\N`. If this theme is general enough, it may be a good idea to mention it higher up in the documentation hierarchy.

What do you think?
